be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/DbActor.scala

```diff
@@ -137,157 +181,249 @@ object DbActor extends AskPatternConstants {
    */
   def apply(mediatorRef: ActorRef, DATABASE_FOLDER: String = "database"): DbActor = {
 
-    val CHANNELS_FOLDER: String = s"$DATABASE_FOLDER/channels"
-
     /* Create a logger for this DB actor"*/
     val logger = LoggerFactory.getLogger("DBLogger")
     
     val options: Options = new Options()
     options.createIfMissing(true)
-    val channelNamesDb: DB =
-      try { factory.open(new File(CHANNELS_FOLDER), options) }
+    
+    val db: DB =
+      try { factory.open(new File(DATABASE_FOLDER), options) }
       catch {
         case e: IOException => {
-          logger.error("Could not open channels folder {}", CHANNELS_FOLDER)
+          logger.error("Could not open database folder {}", DATABASE_FOLDER)
           throw e
         }
 
       }
-    val iterator: DBIterator = channelNamesDb.iterator
-    val initialChannelsMap: mutable.Map[Channel, DB] = mutable.Map.empty
-
-    iterator.seekToFirst()
-    options.createIfMissing(false)
-
-    while (iterator.hasNext) {
-      // open each db associated with each channel name
-      val channelName: String =
-        new String(iterator.next().getKey, StandardCharsets.UTF_8)
-      val channelDb: DB =
-        try { factory.open(new File(channelName), options) }
-        catch {
-          case e: IOException => {
-            logger.error("Could not open channel {}", channelName)
-            throw e
-          }
-
-        }
+    DbActor(mediatorRef, db, DATABASE_FOLDER)
+  }
 
-      // store the channel name and its database in the map
-      initialChannelsMap += (Channel(
-        channelName.toString.stripPrefix(DATABASE_FOLDER)
-      ) -> channelDb)
+  // In the arguments, we might remove the initialChannelsMap and replace the channelNamesDb by the general Db??
+  // FIXME: find way to keep count of channel nb. Idea: store key "channelCount" with value int.getBytes to db and increment it with each channel created
+  sealed case class DbActor(mediatorRef: ActorRef, db: DB, DATABASE_FOLDER: String) extends Actor with ActorLogging {
+    
+    //helper function to determine if a channel should be created during write operations
+    private def writeCreateNewChannel(channel: Channel, message: Message): DbActorMessage = {
```

Comment:
Doesn't this implementation mean that I can create a channel by casting a vote for a non-existent election?

Reply:
The thing is that in the original DbActorSuite, it was a requirement for write to create a new channel when the current one did not exist. However, I indeed find this a little bit weird. However, in the new CastVoteElectionValidator, it will be made sure that only messages to valid election channels will go through, but it certainly does not feel much like defensive programming. We could maybe discuss this a bit more on Slack (I'll send a message there right away), but the current implementation is the same as the previous one.